lib/logout.js

```
@@ -0,0 +1,350 @@
+            var spData = {
+              parsedRequest: {
+                id: requestData.id,
+                serviceProviderLogoutURL: session.serviceProviderLogoutURL
+              },
+              relayState: req.query.RelayState || (req.body && req.body.RelayState)
+            };
+
+            options.store.save(req, spData, function (err, transactionId) {
+              if (err) { return next(err); }
+
+              // We remove the session from the LogoutRequest Originator.
+              // This session is already saved in the store.
+              // We should not send a LogoutRequest to that session
+              // Only a LogoutResponse when there are no other session participants active
+              options.sessionParticipants.remove(requestData.issuer, function (err) {
+                if (err) { return next(err); }
+
+                prepareAndSendLogoutRequest(options.sessionParticipants, transactionId, req, res, next);
+              });
+            });
+          });
+        });
+
+      // Logout flow in progress, incoming SAMLResponse from SP. (Could be SP initiated or IdP initiated)
+      } else if (req.query.SAMLResponse || req.body.SAMLResponse) {
+        function processLogoutResponse(logoutResponse, transactionId) {
+          // LogoutResponse was OK, we remove the session participant from the IdP
+          options.sessionParticipants.remove(logoutResponse.issuer, function (err) {
+            if(err) return next(err);
+
+            // Continue with next session if any
+            prepareAndSendLogoutRequest(options.sessionParticipants, transactionId, req, res, next);
+          });
+        }
+
+        // Verify that the state sent to the SP matches the one returned
+        var h = req.query.RelayState || (req.body && req.body.RelayState);
+        options.store.load(req, h, {destroy: true}, function (err, state) {
+          if (err) { return next(err); }
+          if (!state) { return next(state); }
```

I think we should return an error `if (!state)` (checking)